daemon/src/db.rs

```diff
@@ -463,6 +1053,276 @@ async fn load_cfd_events(
+            vout,
+            payout,
+            price
+        )
+        VALUES
+        (
+            (SELECT id FROM closed_cfds WHERE closed_cfds.uuid = $1),
+            $2, $3, $4, $5
+        )
+        "#,
+        id,
+        cet_txid,
+        cet_payout,
+        cet_vout,
+        price,
+    )
+    .execute(&mut *conn)
+    .await?;
+
+    if query_result.rows_affected() != 1 {
+        anyhow::bail!("failed to insert into cets");
+    }
+
+    Ok(())
+}
+
+async fn insert_refund_tx(
+    conn: &mut Transaction<'_, Sqlite>,
+    id: OrderId,
+    Refund { txid, vout, payout }: Refund,
+) -> Result<()> {
+    let query_result = sqlx::query!(
+        r#"
+        INSERT INTO refund_txs
+        (
+            cfd_id,
+            txid,
+            vout,
+            payout
+        )
+        VALUES
+        (
+            (SELECT id FROM closed_cfds WHERE closed_cfds.uuid = $1),
+            $2, $3, $4
+        )
+        "#,
+        id,
+        txid,
+        vout,
+        payout,
+    )
+    .execute(&mut *conn)
+    .await?;
+
+    if query_result.rows_affected() != 1 {
+        anyhow::bail!("failed to insert into refund_txs");
+    }
+
+    Ok(())
+}
+
+async fn insert_event_log(
+    conn: &mut Transaction<'_, Sqlite>,
+    id: OrderId,
+    events: &[CfdEvent],
+) -> Result<()> {
+    for event in events.iter() {
+        let name = event.event.to_string();
+        let created_at = event.timestamp;
+
+        let query_result = sqlx::query!(
+            r#"
+            INSERT INTO event_log (
+                cfd_id,
+                name,
+                created_at
+            )
+            VALUES
+            (
+                (SELECT id FROM closed_cfds WHERE closed_cfds.uuid = $1),
+                $2, $3
+            )
+            "#,
+            id,
+            name,
+            created_at
+        )
+        .execute(&mut *conn)
+        .await?;
+
+        if query_result.rows_affected() != 1 {
+            anyhow::bail!("failed to insert into event_log");
+        }
+    }
+
+    Ok(())
+}
+
+async fn delete_from_cfds_table(conn: &mut Transaction<'_, Sqlite>, id: OrderId) -> Result<()> {
+    let query_result = sqlx::query!(
+        r#"
+        DELETE FROM
+            cfds
+        WHERE
+            uuid = $1
+        "#,
+        id,
+    )
+    .execute(&mut *conn)
+    .await?;
+
+    if query_result.rows_affected() != 1 {
+        anyhow::bail!("failed to delete from cfds");
+    }
+
+    Ok(())
+}
+
+async fn delete_from_events_table(conn: &mut Transaction<'_, Sqlite>, id: OrderId) -> Result<()> {
+    let query_result = sqlx::query!(
+        r#"
+        DELETE FROM
+            events
+        WHERE cfd_id IN
+            (SELECT id FROM cfds WHERE uuid = $1)
+        "#,
+        id,
+    )
+    .execute(&mut *conn)
+    .await?;
+
+    if query_result.rows_affected() < 1 {
+        anyhow::bail!("failed to delete from events");
+    }
+
+    Ok(())
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
I've added a few tests with the last fixup. They're only based on collaborative closed CFDs because it was easiest to get real events for those.

I've added tests that only concern the `db` layer, proving that the data is moved around correctly.

I've also added a test to verify that the only user of the closed CFDs table (i.e. `projection`) can't tell the difference between loading from `cfds` and `closed_cfds`. The test is not perfect for a few reasons: depending on `projection` in `db/tests`; needing to make some things public in `projection`; the `projection::Cfd` type is built differently for both types of CFD, so the comparison has some caveats.

I still think the test is valuable and we can always iterate on the approach.